### PR TITLE
Add speech asynchronous recognize support.

### PR DESCRIPTION
**Pull request by daspecster**
Add support for Speech API long running operations.

### CODE REVIEW

File: google/cloud/speech/client.py [hunk 1]
```diff
@@ -139,44 +192,115 @@ def sync_recognize(self, content, source_uri, encoding, sample_rate,
...
 
-        if language_code is not None:
-            config['languageCode'] = language_code
-        if max_alternatives is not None:
-            config['maxAlternatives'] = max_alternatives
-        if profanity_filter is not None:
-            config['profanityFilter'] = profanity_filter
-        if speech_context is not None:
-            config['speechContext'] = {'phrases': speech_context}
+def _build_request_data(content, source_uri, encoding, sample_rate,
```
Comment (daspecster) [inline google/cloud/speech/client.py]: This method signature is repeated three times. Once here, once in `async_recognize` and once again in `sync_recognize`. If anyone has suggestions on how to DRY this up I'm happy to implement.

File: docs/speech-usage.rst [hunk 2]
```diff
@@ -30,11 +31,41 @@ create an instance of :class:`~google.cloud.speech.client.Client`.
...
+      ...     'FLAC', 16000, max_alternatives=2)
+      >>> retry_count = 100
+      >>> while retry_count > 0 and not operation.complete:
+      ...     retry_count -= 1
+      ...     time.sleep(10)
+      ...     operation.poll()  # API call
+      >>> operation.complete
+      True
+      >>> operation.results[0]['alternatives'][0]['transcript']
+      "how old is the Brooklyn Bridge"
```
Comment (dhermes) [inline docs/speech-usage.rst]: The Python interpreter will use single quotes if there are no single quotes in the content

File: docs/speech-usage.rst [hunk 3]
```diff
@@ -30,11 +31,41 @@ create an instance of :class:`~google.cloud.speech.client.Client`.
...
+      ...     None, 'gs://my-bucket/recording.flac',
+      ...     'FLAC', 16000, max_alternatives=2)
+      >>> retry_count = 100
+      >>> while retry_count > 0 and not operation.complete:
+      ...     retry_count -= 1
+      ...     time.sleep(10)
+      ...     operation.poll()  # API call
+      >>> operation.complete
+      True
+      >>> operation.results[0]['alternatives'][0]['transcript']
```
Comment (dhermes) [inline docs/speech-usage.rst]: 4 `__getitem__`'s isn't very readable. Can you make it clearer what's intended here? Also, is `operation.results` just the raw parsed JSON?

File: google/cloud/speech/client.py [hunk 4]
```diff
@@ -19,30 +19,8 @@
...
-    MULAW = 'MULAW'
-    """MULAW encoding type."""
-
-    AMR = 'AMR'
-    """AMR encoding type."""
-
-    AMR_WB = 'AMR_WB'
-    """AMR_WB encoding type."""
+from google.cloud.speech.encoding import Encoding
+from google.cloud.speech.operation import Operation
```
Comment (dhermes) [inline google/cloud/speech/client.py]: NOTE: You could've done this refactor in a standalone PR and made this PR easier to review.

File: google/cloud/speech/client.py (hunk 1 above)
Comment (dhermes) [inline google/cloud/speech/client.py]: Looks like you've already done a good job of DRY-ing it up

File: google/cloud/speech/operation.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Long running operation representation for Google Speech API"""
+
+from google.cloud._helpers import _rfc3339_to_datetime
+
+
+class Operation(object):
```
Comment (dhermes) [inline google/cloud/speech/operation.py]: Why not subclass the one found in `google.cloud.operation`?

File: unit_tests/speech/_fixtures.py [hunk 6]
```diff
@@ -28,3 +28,45 @@
 SYNC_RECOGNIZE_EMPTY_RESPONSE = {
     'results': [],
 }
+
+ASYNC_RECOGNIZE_RESPONSE = {
+    'name': '123456789'
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 7]
```diff
@@ -28,3 +28,45 @@
...
+}
+
+OPERATION_COMPLETE_RESPONSE = {
+    'name': '123456789',
+    'metadata': {
+        '@type': ('type.googleapis.com/'
+                  'google.cloud.speech.v1beta1.AsyncRecognizeMetadata'),
+        'progressPercent': 100,
+        'startTime': '2016-09-22T17:52:25.536964Z',
+        'lastUpdateTime': '2016-09-22T17:52:27.802902Z'
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 8]
```diff
@@ -28,3 +28,45 @@
...
+    'done': True,
+    'response': {
+        '@type': ('type.googleapis.com/'
+                  'google.cloud.speech.v1beta1.AsyncRecognizeResponse'),
+        'results': [
+            {
+                'alternatives': [
+                    {
+                        'transcript': 'how old is the Brooklyn Bridge',
+                        'confidence': 0.98267895
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 9]
```diff
@@ -28,3 +28,45 @@
...
+                    }
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 10]
```diff
@@ -28,3 +28,45 @@
...
+                ]
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 11]
```diff
@@ -28,3 +28,45 @@
...
+            }
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 12]
```diff
@@ -28,3 +28,45 @@
...
+        ]
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 13]
```diff
@@ -28,3 +28,45 @@
...
+    }
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 14]
```diff
@@ -28,3 +28,45 @@
...
+OPERATION_INCOMPLETE_RESPONSE = {
+    'name': '123456789',
+    'metadata': {
+        '@type': ('type.googleapis.com/'
+                  'google.cloud.speech.v1beta1.AsyncRecognizeMetadata'),
+        'progressPercent': 27,
+        'startTime': '2016-09-22T17:52:25.536964Z',
+        'lastUpdateTime': '2016-09-22T17:52:27.802902Z'
+    },
+    'done': False
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: unit_tests/speech/_fixtures.py [hunk 15]
```diff
@@ -28,3 +28,45 @@
...
+}
+
+OPERATION_INCOMPLETE_RESPONSE = {
+    'name': '123456789',
+    'metadata': {
+        '@type': ('type.googleapis.com/'
+                  'google.cloud.speech.v1beta1.AsyncRecognizeMetadata'),
+        'progressPercent': 27,
+        'startTime': '2016-09-22T17:52:25.536964Z',
+        'lastUpdateTime': '2016-09-22T17:52:27.802902Z'
```
Comment (dhermes) [inline unit_tests/speech/_fixtures.py]: trailing comma

File: google/cloud/speech/connection.py [hunk 16]
```diff
@@ -26,7 +26,7 @@ class Connection(base_connection.JSONConnection):
     API_VERSION = 'v1beta1'
     """The version of the API, used in building the API call's URL."""
 
-    API_URL_TEMPLATE = '{api_base_url}/{api_version}/speech:{path}'
+    API_URL_TEMPLATE = '{api_base_url}/{api_version}/{path}'
```
Comment (dhermes) [inline google/cloud/speech/connection.py]: Why did `speech:` have to come out?

File: docs/speech-usage.rst (hunk 3 above)
Comment (daspecster) [inline docs/speech-usage.rst]: Yeah it is.
I'll clean up the API surface for this if you're cool with that.

`operation.results[0].transcript` and `operations.results[0].confidence` sound ok?

File: docs/speech-usage.rst (hunk 2 above)
Comment (daspecster) [inline docs/speech-usage.rst]: ✅

File: google/cloud/speech/client.py (hunk 4 above)
Comment (daspecster) [inline google/cloud/speech/client.py]: Yeah, I should have planned this out more before jumping into it. Noted for streaming.

File: google/cloud/speech/operation.py (hunk 5 above)
Comment (daspecster) [inline google/cloud/speech/operation.py]: I thought they would diverge more than they did. I'll subclass it in the updates.

File: google/cloud/speech/connection.py (hunk 16 above)
Comment (daspecster) [inline google/cloud/speech/connection.py]: To support the `/operations` endpoint.

1. `/speech:syncrecognize`
2. `/speech:asyncrecognize`
3. `/operations`

 Let me know if I should put it back and add logic to handle `operations` somewhere.

File: unit_tests/speech/test_client.py [hunk 17]
```diff
@@ -172,6 +172,24 @@ def test_sync_recognize_with_empty_results(self):
...
+        client = self._makeOne(credentials=credentials)
+        client.connection = _Connection(RETURNED)
+
+        encoding = Encoding.FLAC
+
+        operation = client.async_recognize(None, self.AUDIO_SOURCE_URI,
+                                           encoding,
+                                           self.SAMPLE_RATE)
+        self.assertIsInstance(operation, Operation)
+        self.assertFalse(operation.complete)
```
Comment (tseaver) [inline unit_tests/speech/test_client.py]: Should there be a `.metadata` instance on the operation?

**Comment by daspecster** [comment]
@dhermes was there supposed to be a link in here?
>I'm not 100% but I think there is a utility in google.protobuf for parsing a JSON version of a protobuf message type (here it is Operation)

File: docs/speech-usage.rst (hunk 3 above)
Comment (tseaver) [inline docs/speech-usage.rst]: Shouldn't `operation.results` map onto the sampe type as is returned from `sync_recognize`?  I.e. one dict per alternative?

File: google/cloud/speech/connection.py (hunk 16 above)
Comment (tseaver) [inline google/cloud/speech/connection.py]: The docs for both [syncrecognize](https://cloud.google.com/speech/reference/rest/v1beta1/speech/syncrecognize) and [asyncrecognize](https://cloud.google.com/speech/reference/rest/v1beta1/speech/asyncrecognize) have the `speech:` prefix.  It looks like @daspecster made it an explicit part of the `api_request` calls below.

File: docs/speech-usage.rst (hunk 3 above)
Comment (daspecster) [inline docs/speech-usage.rst]: I'm not sure how to do away with the `[0][0]` actually. As much as I would like too.

Comment (daspecster) [inline docs/speech-usage.rst]: @tseaver, I can do that just like [`sync_recognize`](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/b7f004e00f1ee021b0cb7ac9261fcc6c021dfacb/speech/google/cloud/speech/client.py#L180) but then we'll have to update it later for batch handling.

File: unit_tests/speech/test_client.py (hunk 17 above)
Comment (daspecster) [inline unit_tests/speech/test_client.py]: I could make a `SpeechMetadata` class if you want. `operation.metadata.last_updated`..etc?

**Comment by daspecster** [comment]
@tseaver I think I got all the things you mentioned. LMKWYT

File: unit_tests/speech/test_operation.py [hunk 18]
```diff
@@ -84,8 +84,10 @@ def test_poll_complete(self):
 class _Connection(object):
     def __init__(self, response):
         self.response = response
+        self._requested = []
 
     def api_request(self, method, path):
+        self._requested.append({'method': method, 'path': path})
```
Comment (tseaver) [inline unit_tests/speech/test_operation.py]: Hmm, doesn't look like the tests are making any assertion about the value of `connection._requested`.

**Comment by tseaver** [comment]
@daspecster Should be OK to merge if you can tweak the assertions for the operations tests (see above).

File: unit_tests/speech/test_operation.py (hunk 18 above)
Comment (daspecster) [inline unit_tests/speech/test_operation.py]: I added a lot for this so you may want to take a look at it again.
